File: apps/common/templatetags/truncate_middle.py
"""Truncate middle template filter."""

from django import template


register = template.Library()


@register.filter
def truncate_middle(text, max_length):
    """Truncate the text in the middle if it is longer than the max_length.

    Args:
        text: The text to truncate.
        max_length: The maximum length of the text.

    Returns:
        The truncated text.

    """
    text = str(text)

    if len(text) <= max_length:
        return text

    # Calculate the number of characters to show before and after the ellipsis
    chars_to_show = max_length - 3
    front_chars = chars_to_show // 2
    back_chars = chars_to_show - front_chars

    return text[:front_chars] + "..." + text[-back_chars:]

File: static_build/js/utils/truncateMiddle.js
export const truncateMiddle = function (text, maxLength) {
    if (text.length <= maxLength) {
        return text;
    }

    // Calculate the number of characters to show before and after the ellipsis
    const charsToShow = maxLength - 3;
    const frontChars = Math.ceil(charsToShow / 2);
    const backChars = Math.floor(charsToShow / 2);

    return text.substr(0, frontChars) + '...' + text.substr(text.length - backChars);
};
